### ISSUE TITLE
v3.5.5 changes the http response status codes

### ISSUE THREAD

**Issue by pmareke** [argoproj/argo-workflows#13558]
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I have tested with the `:latest` image tag (i.e. `quay.io/argoproj/workflow-controller:latest`) and can confirm the issue still exists on `:latest`. If not, I have explained why, **in detail**, in my description below.
- [X] I have searched existing issues and could not find a match for this bug
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened? What did you expect to happen?

In the latest images some endpoints change their status responses.

When a call this endpoints:

- Calling the API with an invalid token.
- Calling the API with a token without enough permissions.

I start returning a 404 when in previous version (3.4.x) the status code was a 401.

Also when:

- Calling the API to stop a non existing workflow.
- Calling the API to resume a non existing workflow.

I stop returning a 404 when in previous version (3.4.x) the status code was a 404.

### Version(s)

v3.5.x

### Paste a minimal workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
It's an http call, not related with the workflow itself.
```

### Logs from the workflow controller

```text
Nothing to show.
```

### Logs from in your workflow's wait container

```text
Nothing to show.
```

**Comment by agilgur5** [comment]
Could you give some example API requests to make this easier to reproduce and debug?

**Comment by pmareke** [comment]
Sure @agilgur5:

 -  GET`/workflows/{self.namespace}/{workflow_name}`sending an invalid Bearer token -> 401 in v3.4.x and 404 in v3.5.x
 - GET `/workflows/{self.namespace}/{workflow_name}`sending aa valid Bearer token without enough permissions -> 401 in v3.4.x and 404 in v3.5.x

It also change the response of the following endpoints:
- PUT `/workflows/{self.namespace}/{workflow_name}/resume` with a valid token and a non existing workflows  -> 404 in v3.4.x
- PUT `/workflows/{self.namespace}/{workflow_name}/stop` with a valid token and a non existing workflows -> 404 in v3.4.x

**Comment by agilgur5** [comment]
Some more follow-up questions:

1. Are the first two also non-existent Workflows?
1. Do the namespaces exist?
1. What is the response to the last two? A 200?
1. Also what auth mode(s) are you using?

**Comment by pmareke** [comment]
> Some more follow-up questions:
> 
> 1. Are the first two also non-existent Workflows? `NO`
> 2. Do the namespaces exist? `YES`
> 3. What is the response to the last two? A 200? `Not sure, I could't see, I need to rollback the upgrade 😞.`
> 4. Also what auth mode(s) are you using? `The Authorization header with the Bearer.`

The most important info I can give you is that without changing nothing on my side a lot of tests, which validates the ArgoWorkflow API responses, start to fail.

Rolling back to the v3.4 version everything back to work again.

**Comment by agilgur5** [comment]
> NO

Ok, so it's giving a 404 instead of a 401 even when the Workflow exists 🤔 

> Not sure, I could't see, I need to rollback the upgrade as fast as possible 😞.

I imagine you were able to test locally or on a staging/dev env if you tried `:latest`, so that's where you could repro.

> The Authorization header with the Bearer token.

There's two of those; `client` is `Bearer ****` and `sso` is `Bearer v2:****`
Also you can have multiple enabled, so using one does not preclude the other. That's why I asked, as a response from one auth method could be different in another auth method (see also https://github.com/argoproj/argo-workflows/pull/13372 et al)

> Rolling back to the v3.4 version everything back to work again.

Yea I got that, and I did mark this as a regression according to that, I'm just not recalling any _intentional_ change there, so it's likely an unintentional bug somewhere, which is a lot harder to track down 🫠

**Comment by Joibel** [comment]
My best guess would be this is an unintentional side effect of #13021.

**Comment by agilgur5** [comment]
That's v3.5.8+ and only affects the list endpoint, vs this is happening on individual workflows, and even on stops. So I would think #11121 would have caused it rather

**Comment by pmareke** [comment]
> > NO
> 
> Ok, so it's giving a 404 instead of a 401 even when the Workflow exists 🤔

YES

> > Not sure, I could't see, I need to rollback the upgrade as fast as possible 😞.
> 
> I imagine you were able to test locally or on a staging/dev env if you tried `:latest`, so that's where you could repro.

Yes, sure. I'll update again run the tests and give you more info.

> > The Authorization header with the Bearer token.
> 
> There's two of those; `client` is `Bearer ****` and `sso` is `Bearer v2:****` Also you can have multiple enabled, so using one does not preclude the other. That's why I asked, as a response from one auth method could be different in another auth method (see also #13372 et al). In particular, `server` is effectively no auth (if your Server has appropriate RBAC)

In my case I'm using the  `Bearer ****` one.

> > Rolling back to the v3.4 version everything back to work again.
> 
> Yea I got that, and I did mark this as a regression according to that, I'm just not recalling any _intentional_ change to API responses, so it's likely an unintentional bug somewhere, which is a lot harder to track down 🫠 More information helps with a repro to try to debug that

Yes, I was taking a look to the latest releases notes and I didn't see too any change related 😔.

**Comment by pmareke** [comment]
Ok so in the following endpoints:
`PUT /workflows/{self.namespace}/{workflow_name}/resume` with a valid token and a non existing workflow.
`PUT /workflows/{self.namespace}/{workflow_name}/stop` with a valid token and a non existing workflow.

The response now does not include the workflow name as it did in previous versions:

 <img width="642" alt="image" src="https://github.com/user-attachments/assets/6cc900c8-6d1f-4bb3-8230-53b76f7fdc7a">

But I don't think this is nothing to worry about, my test are validating the text error and that's an implementation detail and I need to change it.

The changes then are related with:

GET/workflows/{self.namespace}/{workflow_name} sending an invalid Bearer token -> 401 in v3.4.x and 404 in v3.5.x
GET /workflows/{self.namespace}/{workflow_name} sending a valid Bearer token without enough permissions -> 401 in v3.4.x and 404 in v3.5.x

**Comment by agilgur5** [comment]
> I have the feeling that maybe the changes comes from the **cli**, does it makes any sense to you?

No. The CLI doesn't affect responses from the API, it's just a client.

**Comment by agilgur5** [comment]
Ok pretty sure I found the root cause, and my intuition was roughly correct. There was a precursor to #11121, which was https://github.com/argoproj/argo-workflows/pull/11055 and #11674. I'll follow-up more tomorrow and this week, there might be some other bugs related to that as well.

**Comment by agilgur5** [comment]
Correction, #11674 was correct, but #11121 indeed made some alterations to the same exact place apparently, [these lines](https://github.com/argoproj/argo-workflows/blob/ef09d9ffed0fb3c0f3e672db588d6a7c6cf2a5f5/server/workflow/workflow_server.go#L721-L723) specifically.

I should have a fix out soon

**Comment by salvamadrid** [comment]
Any news about that?

**Comment by Joibel** [comment]
This was fixed by #13957 ("fix(api): properly authorize GET workflow fallback to archive"), which shipped in v3.5.13, v3.6.2 and v3.7.0. It went out as part of GHSA-h36c-m3rf-34h9, which is why it was never linked back to this issue.

The cause was the archive fallback in `getWorkflow()`: when the live Kubernetes `Get` failed with 401/403, the archive lookup returned no workflow and no error, and the code then returned a bare `codes.NotFound` - masking the real auth error as a 404. That branch is gone. Every failure path now goes through `getWorkflowOrigErr()`, which preserves the original error, so an invalid token gives 401 and an under-privileged token gives 403. There are e2e assertions covering both in `test/e2e/argo_server_test.go` (`GetWFsFallbackArchivedFakeToken` -> 401, `GetWFsFallbackArchivedBadToken` -> 403).

The same change also fixes the second symptom: the resume/stop responses once again carry the Kubernetes "not found" message including the workflow name, rather than a bare "not found".

One thing to flag: a valid token that lacks permission now returns 403, not the 401 you saw on v3.4.x. That matches what the Kubernetes API returns for insufficient RBAC, so it is the intended code.

Closing as fixed - please reopen or file a fresh issue if you still see a 404 on v3.5.13+ / v3.6.2+ / v3.7.x.

**Comment by pmareke** [comment]
Great, thanks @Joibel
